### DISCUSSION TITLE
Access User Information

### DISCUSSION THREAD

**Discussion by queza85** [exoplatform/exo-community#49] [Q&A]
Hi 
Finally, we are success to connect user with Keycloak.
And login success via keycloak user.
But we found that the user information need to be pushed exo db(mysql).

if we have a new user in keycloak, we should put in this information to mysql directly, right?

I have some question to handle the user information directly in db but it didn't update even restart docker server

1. Modify User information
- I just added new field for user information in jbid_io_attr and jbid_io_attr_text_values under the same user id. but it can not see in user profile exo site. so do I need to add extra table? as my understanding is correct, user information is all in the jbid_* tables.
- I tested to change existed field like firstName but it didn't update it in user profile too

2. Handle user the permission
- I think jbid_io can manage account permission based on jbid_io_type. 
  So I changed type from USER to root_type. after that I login using issued ID, it has 404 error.
  It needs to be changed another table?

Always thanks for your advise.

**Comment by rdenarie** [comment]
Hi @queza85,

`if we have a new user in keycloak, we should put in this information to mysql directly, right?`
In general manner, it is not q good idea to launch queries on database directly. It will certainly generates problems with Hibernate object we use.

For 1) Which kind of information would you put in eXo ?
In fact, there are 2 tables for profile in eXo. And old (historical one) with table jbid_io, and the one displayed in eXo in SOC_IDENTIT* tables. 
When an information is put in one of this profile, a listener is launch to push the info into the other one.
But, as you change in database, the change is not detected by the code, and the listener not launched.
What is your user database ? and LDAP ? Active Directory ? other ?

For 2) jbid_io_type is not used for permissions. You shoulnt change it.
Which kind of permission you want to add ? Where these permissions are stored ?

**Reply by queza85** [reply]
Hi @rdenarie 

1) We use AD. does it anyway to notification to exo when AD hsa changed?

2) I mean the role for user. for example, when I add user, I would like to give a role as ROOT.

**Reply by rdenarie** [reply]
For LDAP/AD integration, you will have to configure eXo to read your ad. Then, user properties will be automatically synced : https://docs.exoplatform.org/administration/LDAP.html

The "root" role, is beiing in group /platform/administrator. 
So, with "root" user, you can manually add your users in this group. 

With AD configuration, you will be also able to see in eXo your AD groups. 
BUT, you will not be able to automatically put a user in administrators group.

**Reply by queza85** [reply]
thanks for your information.

If we use some custom role like capture, we cannot add this role exo db directly?

![image](https://github.com/exoplatform/exo-community/assets/2861131/d76ab6bc-d5fa-4763-89e7-cf17b1093234)

**Reply by rdenarie** [reply]
No, it is not possible directly in db. 
Can you explain your functionnal need ? so that I can search a solution to your request

**Reply by queza85** [reply]
We have multiple AD. It connected this with eXo via Keycloak.
So that is why I pushed user to exo mysql directly.
I know this is not good way as you mentioned, but I cannot find any good way to use multiple AD.

1. cannot login generated user
    I added SOC_IDENTITIES, SOC_IDENTITY_PROPERTIES to mysql directly.
    But it cannot be login with 403 error even we restarted server.
    So I added jbid_io, jbid_io_attr and jbid_io_attr_text_values also using same data but it cannot login too.

2. which table can I add the role?
  If I generated user in table, I would like to assign custom role like above capture to generated user.
  I thought I can add some information related table in the exo mysql. 
  If it is cannot added, how to manage user role inside mysql? it has no table for this?

**Reply by rdenarie** [reply]
Hello,
I think there are some misunderstood on about eXo work, and on what exo can do or not.

Changing thigs directly in DB is not the good way to do, and could generate problem to use the platform. 

eXo IS NOT able to use multiple AD. We need a single AD.

Options are : 

- Use IDP as Keycloack. When the user logs in, it is automatically created in eXo. BUT you cannot "autofill" profile with this method
- Use something to mirror and centralize all your AD. Then you will be able to plug eXo to this centralized user database. With this, you will be able to have all your users in eXo directly, and you will be able to sync profile attributes. In addition, you will be able to synchronize groups and roles from yours AD


I will not give you information about how update database because I have no information about it. We only use Hibernate to change database information.

Perhaps, it could be better to use REST API (https://docs.exoplatform.org/guide/openapi/social.html#/v1%2Fsocial%2Fusers/updateUserProfileAttributes) if you want to update user profile

**Reply by queza85** [reply]
Hi @rdenarie 

We use IDP as Keycloak. So I expected that I login using keycloak user. 
But it has error 403 after authorise by keycloak login and then return to exo.
It looks new user didn't create in exo db.
So that is why I start to put in user information to db manually. 

As follow your first option, it should login even that is new user in keycloak side. 
Do I need to check any option for this?

**Reply by queza85** [reply]
hi this is solved when I synchronise all the fields in the db. thanks!
